src/pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -61,13 +61,16 @@ def __init__(self, trainer: "pl.Trainer", multiple_trainloader_mode: str = "max_
     def _should_reload_train_dl(self) -> bool:
         """Check if train dataloader should be reloaded."""
         n_epochs = self.trainer.reload_dataloaders_every_n_epochs
-        return n_epochs and (self.trainer.current_epoch - self.trainer._last_train_dl_reload_epoch >= n_epochs)
+        return n_epochs and (self.trainer.current_epoch % n_epochs == 0)
 
     @property
     def _should_reload_val_dl(self) -> bool:
         """Check if validation dataloader should be reloaded."""
         n_epochs = self.trainer.reload_dataloaders_every_n_epochs
-        return n_epochs and (self.trainer.current_epoch - self.trainer._last_val_dl_reload_epoch >= n_epochs)
```

Comment:
> But this PR wants to reload also at the very beginning regardless

not really. It depends on the sanity check and `check_val_every_n_epoch`.

so everything is decided when `trainer.reset_val_dataloader` is called. This is called either during sanity check or during validation.

If a sanity check is happening:
- validation is enabled for the same epoch (in an ideal case 0), then `self.trainer._last_val_dl_reload_epoch` here will not be None here, and we don't have to reload the dataloaders again for that epoch.
- validation is not enabled for that epoch, then `self.trainer._last_val_dl_reload_epoch` will be None here during first validation and we have to reload the dataloaders regardless so once they are loaded, `_last_val_dl_reload_epoch` will get update inside `trainer.reset_val_dataloader`.

if sanity check is not happening, then this will be triggered only when `check_val_every_n_epoch` is satisfied, and we have to reload the dataloaders regardless, and again this will be updated during `trainer.reset_val_dataloader` for the next time.